### PR TITLE
Consumer: revamp message tracking

### PR DESCRIPTION
**Pull request by Annopaolo**
> _"Se vogliamo che tutto rimanga com’è, bisogna che tutto cambi”_

(Tancredi in "Il Gattopardo", by G. Tomasi di Lampedusa)

### What
Essentially, revisit the internals of the consumer application to remove old and messy code. The new supervision structure brought forward in #6 was the first step, and now the actual implementation of AMQPDataConsumer, MessageTracker and DataUpdater is simplified.

At a first glance, the most relevant change is that it is the MessageTracker (and not the AMQPDataConsumer) tasked with ack-/reject-ing a message, once processing is done. This allows to remove a level of indirection and communication between processes, making the system more reliable and fault-tolerant, because the application state is less distributed.

### Message flow
The standard message flow is now as follows:
mississippi_flow drawio

This allows the MessageTracker to maintain the ordering invariant. Internally, it does so by keeping a FIFO queue of messages in its state, and checking that only the message on top of the queue is being processed at any given time.

There exist only one AMQPDataConsumer per queue, and only one MessageTracker and DataUpdater per sharding key. This allows for an AMQPDataConsumer to forward messages to more than one MessageTracker, but a given MessageTracker can receive messages from a single AMQPDataConsumer only. Moreover, only one MessageTracker and one DataUpdater exist for a given sharding key at any given time.

### Fault handling
The supervision tree is now as follows (see #6 for more)
mississippi_consumer_supervision_tree drawio

- The AMQPDataConsumer monitors each MessageTracker to which it is sending messages, so that if they crash, it crashes too and messages are put back in the AMQP queues
- The AMQPDataConsumer is linked to the AMQP Channel it receives messages from so that in case of a channel crash, it crashes too and messages are put back in the AMQP queues
- The MessageTracker monitors the DataUpdater to which it is sending messages, so that if it crashes, it can request another DataUpdater process for the same key and let the new process handle the message
- The MessageTracker monitors the AMQP Channel from which it is receiving (via AMQPDataConsumer) messages, so that if the channel crashes, it crashes too and messages are put back in the AMQP queues

### CODE REVIEW

**Review by noaccOS — COMMENTED** [review]
IMO AMQPDataConsumer should not be a process.
- All it does as a process is relay messages to MessageTracker. MessageTracker could receive messages from the channel directly
- The other functionality (such as queue setup) could be done as a library
- We already have a self-contained process which crashes in case of channel errors: the channel
- AMQPDataConsumer is effectively linked to the channel process. The same observability could be achieved monitoring the channel pid from the MessageTracker 

TL;DR: to me it looks like an abstraction for the channel process when all it really needed was a library to handle the initialization

**Comment by Annopaolo** [comment]
> TL;DR: to me it looks like an abstraction for the channel process when all it really needed was a library to handle the initialization

@noaccOS there is a significant difference between AMQPDataConsumer and MessageTracker: while the former fetches messages from a queue (say `mississippi_data_1`), the latter handle only messages for a given sharding key (say `shard_x`). As messages from more than one shard might be put in the same queue (e.g. `shard_x` and `shard_y` both are hashed to `1`), there exists the need for a process able to retrieve messages from that queue and pass them to the right tracker.
A further point in favour of keeping the consumer and tracker processes separate is the cardinality: while the number of consumers is bounded (and fixed: it is always equal to  `queue_total_count`), the number of possible sharding keys is not, as they can be any erlang term. Trackers can also be inactive at a given point in time, and spawned (activated) when needed with `MessageTracker.get_message_tracker/1`.
Finally, note that there is no way of knowing the sharding key of a message statically from just the queue number, as the queue number  is an hash of the key. In order to retrieve it, you have to dequeue the message and read the `sharding_key` header.
You are right on making the MessageTracker state slimmer: we don't need the queue count anymore, as the sharding key value is now in the process

File: lib/consumer/message_tracker/server.ex [hunk 1]
```diff
@@ -1,222 +1,159 @@
...
+        new_state = drop_head_from_queue(state)
+        # let's move on to the next message
+        {:reply, :ok, new_state, {:continue, :process_message}}
 
-    :ok = ack(acknowledger, delivery_tag)
-
-    {:reply, :ok, {:accepting, new_queue, new_ids, acknowledger}}
+      _ ->
+        # discard the message
+        {:reply, :ok, state, {:continue, :process_message}}
```
[the remark this answers is not in the document]
Comment (Annopaolo) [reply lib/consumer/message_tracker/server.ex:55]: We want to continue to `:process_message` in order to move on with the next message in the queue. Maybe it would be clearer if the state was called `:next_message`?
The fact that the message received via `ack_received` is not acked is enough to discard it, as we do nothing more with it

File: lib/consumer/data_updater.ex [hunk 2]
```diff
@@ -150,39 +105,35 @@ defmodule Mississippi.Consumer.DataUpdater do
...
+  def handle_cast({:handle_message, %Message{} = message}, state) do
+    %Message{payload: payload, headers: headers, timestamp: timestamp, meta: meta} = message
+
+    case state.message_handler.handle_message(
+           payload,
+           headers,
+           meta.message_id,
+           timestamp,
+           state.handler_state
+         ) do
```
[the remark this answers is not in the document]
Comment (Annopaolo) [reply lib/consumer/data_updater.ex:117]: Makes sense! It would probably lead to a destructuring pattern match in the library client's code, but it's cleaner

File: lib/consumer/amqp_data_consumer/state.ex [hunk 3]
```diff
@@ -0,0 +1,10 @@
+defmodule Mississippi.Consumer.AMQPDataConsumer.State do
+  use TypedStruct
+
+  typedstruct do
+    field :queue_name, String.t(), enforce: true
+    field :queue_range, Range.t(), enforce: true
+    field :queue_total_count, pos_integer(), enforce: true
```
[the remark this answers is not in the document]
Comment (Annopaolo) [reply lib/consumer/amqp_data_consumer/state.ex:7]: We actually don't need neither one nor the other!

File: lib/consumer/data_updater.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (Annopaolo) [reply lib/consumer/data_updater.ex:117]: Yes, let's move it outside this PR, it's quite big already

**Review by eddbbt — APPROVED** [review]
